### PR TITLE
Fix Matthews correlation coefficient when the denominator is 0

### LINKED ISSUE #779: Matthews correlation coefficient should return 0 instead of NaN
## 🐛 Bug

By [definition](https://en.wikipedia.org/wiki/Phi_coefficient), "if any of the four sums in the denominator is zero, the denominator can be arbitrarily set to one; this results in a Matthews correlation coefficient of zero, which can be shown to be the correct limiting value."
However, the [code](https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/functional/classification/matthews_corrcoef.py#L22) does not check for this case and will return NaN.

### LINKED ISSUE #780: Add `RetrievalRecallAtFixedPrecision`
## 🚀 Feature

Add a new metric calculating (retrieval) recall at K at fixed precision  (typically `K=1` for this case).

### Motivation

From my experience in IR, there are use cases when a user should be provided with some results on top position (e.g. featured snippets in google search) only with some precision threshold. Subsequently, one may be interested in the recall to know which portion of queries will be handled.

### Pitch

Add a new retrieval metric `RetrievalRecallAtFixedPrecision`, which is inspired by `BinnedRecallAtFixedPrecision` and `RetrievalPrecision`.

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #779

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: torchmetrics/functional/classification/matthews_corrcoef.py [hunk 1]
```diff
@@ -37,7 +37,15 @@ def _matthews_corrcoef_compute(confmat: Tensor) -> Tensor:
...
     c = torch.trace(confmat).float()
     s = confmat.sum().float()
-    return (c * s - sum(tk * pk)) / (torch.sqrt(s ** 2 - sum(pk * pk)) * torch.sqrt(s ** 2 - sum(tk * tk)))
+
+    cov_ytyp = c * s - sum(tk * pk)
+    cov_ypyp = s ** 2 - sum(pk * pk)
+    cov_ytyt = s ** 2 - sum(tk * tk)
+
+    if cov_ypyp * cov_ytyt == 0:
+        return 0.0
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/matthews_corrcoef.py:46]:
```suggestion
        return torch.tensor(0, dtype=confmat.dtype, device=confmat.device)
```

**Review by stancld — APPROVED** [review]
LGTM!
